File: packages/sdk/src/audit/chain.ts
/**
 * Cryptographic audit chain for append-only tamper detection.
 *
 * Each record's hash is computed over the previous hash + the record's
 * deterministic JSON serialization, forming a hash chain. Any mutation
 * to a historical record invalidates all subsequent hashes.
 *
 * @module audit/chain
 */

import { createHash } from 'node:crypto';

/**
 * Deterministic JSON serialization with sorted object keys.
 * Required for stable hashes across V8 versions and key insertion orders.
 */
function sortedStringify(obj: unknown, seen = new WeakSet<object>()): string {
	if (obj === null || typeof obj !== 'object') return JSON.stringify(obj);
	if (seen.has(obj)) return '"[Circular]"';
	seen.add(obj);
	if (Array.isArray(obj)) return '[' + obj.map((v) => sortedStringify(v, seen)).join(',') + ']';
	const keys = Object.keys(obj as Record<string, unknown>).sort();
	return '{' + keys.map((k) =>
		JSON.stringify(k) + ':' + sortedStringify((obj as Record<string, unknown>)[k], seen),
	).join(',') + '}';
}

export function computeChainHash(prevHash: string, record: unknown): string {
	const payload = prevHash + sortedStringify(record);
	return createHash('sha256').update(payload).digest('hex');
}

/** Genesis hash — used as prevHash for the very first record */
export const GENESIS_HASH = '';

File: packages/sdk-python/veto/audit/chain.py
"""
Cryptographic audit chain for append-only tamper detection.

Each record's hash is computed over the previous hash + the record's
deterministic JSON serialization, forming a hash chain. Any mutation
to a historical record invalidates all subsequent hashes.
"""

import hashlib
import json
from typing import Any

GENESIS_HASH = ""


def sorted_stringify(obj: Any, seen: set[int] | None = None) -> str:
    """Deterministic JSON serialization with sorted keys."""
    if seen is None:
        seen = set()
    if obj is None:
        return "null"
    if isinstance(obj, bool):
        return "true" if obj else "false"
    if isinstance(obj, (int, float)):
        return json.dumps(obj)
    if isinstance(obj, str):
        return json.dumps(obj)

    obj_id = id(obj)
    if obj_id in seen:
        return '"[Circular]"'
    seen.add(obj_id)

    if isinstance(obj, list):
        return "[" + ",".join(sorted_stringify(v, seen) for v in obj) + "]"
    if isinstance(obj, dict):
        keys = sorted(obj.keys())
        return (
            "{"
            + ",".join(
                json.dumps(k) + ":" + sorted_stringify(obj[k], seen)
                for k in keys
            )
            + "}"
        )
    return json.dumps(str(obj))


def compute_chain_hash(prev_hash: str, record: Any) -> str:
    """SHA256 hash of prev_hash + sorted_stringify(record)."""
    payload = prev_hash + sorted_stringify(record)
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()
